cloudferrylib/os/storage/cinder_database.py

```diff
@@ -22,23 +23,106 @@
 
 CINDER_VOLUME = "cinder-volume"
 LOG = utl.get_log(__name__)
+ID = 'id'
+DISPLAY_NAME = 'display_name'
 PROJECT_ID = 'project_id'
 STATUS = 'status'
 TENANT_ID = 'tenant_id'
 USER_ID = 'user_id'
 DELETED = 'deleted'
 HOST = 'host'
-IGNORED_TBL_LIST = ('quota_usages')
+IGNORED_TBL_LIST = ('quotas', 'quota_usages')
 QUOTA_TABLES = (
     'quotas',
     'quota_classes',
     'quota_usages',
 )
 
+# SKIPPING_STATUSES = ['creating', 'error', 'error_deleting',
+# 'error_attaching']
+VALID_STATUSES = ['available', 'in-use', 'attaching', 'detaching']
+
+
+def skip_invalid_status_volumes(volumes):
+    """Filter volumes with valid status.
+
+    :return: list
+
+    """
+    result = []
+    for v in volumes:
+        if v[STATUS] not in VALID_STATUSES:
+            LOG.error('Skipping volume %s[%s] in "%s" state',
+                      v.get(DISPLAY_NAME, ''), v[ID], v[STATUS])
+        else:
+            result.append(v)
+    return result
+
+
+def get_key_and_auto_increment(cursor, table):
+    """ Get name of column that is primary_key, get auto_increment.
+
+    :return: (primary_key, auto_increment)
+
+    """
+    cursor.execute(
+        "show index from {table} where Key_name = 'PRIMARY'".format(
+            table=table))
+    primary_key = cursor.fetchone().get("Column_name")
+    cursor.execute(
+        "select auto_increment from information_schema.tables"
+        " where table_name = '{table}'".format(table=table))
+    auto_increment = cursor.fetchone().get("auto_increment")
+    return primary_key, auto_increment
+
+
+def filter_data(existing_data, data_to_be_added, primary_key,
```

Comment:
Logic looks quite complex here, would be good to have it covered with unit tests